README.md
[![Coffee IT - Aroma CITCardsCarousel iOS Component](https://coffeeit.nl/wp-content/uploads/2022/10/Aroma-Card-Carousel-cover-iOS-1.png)](https://coffeeit.nl/)

[![Swift](https://img.shields.io/badge/Swift-5.6-red?style=flat-square)](https://img.shields.io/badge/Swift-5.6-red?style=flat-square)
[![iOS 14+](https://img.shields.io/badge/iOS-v14+-pink?style=flat-square)](https://img.shields.io/badge/iOS-v14+-pink?style=flat-square)
[![Mirror Repository](https://img.shields.io/badge/Mirror-Repository-pink?style=flat-square)](https://img.shields.io/badge/Mirror-Repository-pink?style=flat-square)
[![License](https://img.shields.io/badge/License-MIT-pink.svg?style=flat-square)](LICENSE.md)
[![Swift Package Manager](https://img.shields.io/badge/Swift_Package_Manager-Compatible-red?style=flat-square)](https://img.shields.io/badge/Swift_Package_Manager-Compatible-red?style=flat-square)
[![LinkedIn](https://img.shields.io/badge/LinkedIn-@CoffeeIT-blue.svg?style=flat-square)](https://linkedin.com/company/coffee-it)
[![Facebook](https://img.shields.io/badge/Facebook-CoffeeITNL-blue.svg?style=flat-square)](https://www.facebook.com/CoffeeITNL/)
[![Instagram](https://img.shields.io/badge/Instagram-CoffeeITNL-blue.svg?style=flat-square)](https://www.instagram.com/coffeeitnl/)
[![Twitter](https://img.shields.io/badge/Twitter-CoffeeITNL-blue.svg?style=flat-square)](https://twitter.com/coffeeitnl)

The CITCardsCarousel package provides a configurable arrangement of swipe-able card views which can be used for tutorials and other flows.
Includes page indicators, skip, previous, next and finish buttons.

<p float="left">
    <img src="Docs/Images/cards-carousel.gif" width="300">
</p>

## ⚡ Installation

### 🔨 SwiftPM

To install the Swift Package, go to Project > Package Dependencies > + > Search or Enter Package URL > Fill in:
```
https://github.com/Coffee-IT-Development/CardsCarousel-iOS-Component
```

## 📖 Usage

Import CITCardsCarousel and add a CITCardsCarouselView to your SwiftUI view.
The following example showcases how you could use the cards carousel in a sheet or fullScreenCover.

```swift
import CITCardsCarousel
import SwiftUI

struct CITCardsCarouselExampleView: View {
    @State private var selectedTab = 0
    @State private var isShowingInSheet = false
    @State private var isShowingFullscreen = false
    
    var body: some View {
        VStack {
            exampleButton("Show cards carousel fullscreen", action: showCardsCarouselFullscreen)
            exampleButton("Show cards carousel in sheet", action: showCardsCarouselInSheet)
        }
        .fullScreenCover(isPresented: $isShowingFullscreen) {
            cardsCarousel
        }
        .sheet(isPresented: $isShowingInSheet) {
            cardsCarousel
        }
    }
    
    var cardsCarousel: some View {
        CITCardsCarouselView(selection: $selectedTab, pageCount: 6, config: .coloredExample) {
            card("A").tag(0)
            card("B").tag(1)
            card("C").tag(2)
            card("D").tag(3)
            card("E").tag(4)
            card("F").tag(5)
        }
    }

    private func card(_ name: String) -> some View {
        ZStack {
            Color.white
            Text(name)
        }
    }
    
    private func exampleButton(_ text: String, action: @escaping () -> Void) -> some View {
        Button(action: action) {
            Text(text)
                .padding()
                .foregroundColor(.carouselButtonTextColor)
                .background(Color.coffeeItColor)
                .cornerRadius(20)
        }
    }
    
    private func showCardsCarouselFullscreen() {
        selectedTab = 0
        isShowingFullscreen = true
    }
    
    private func showCardsCarouselInSheet() {
        selectedTab = 0
        isShowingInSheet = true
    }
}
```

## ⚙️ Customization

```swift
/// The padding around the card container. Defaults to `.init(top: 24, leading: 24, bottom: 32, trailing: 24)`.
public var cardPadding: EdgeInsets

/// The safe area edges that a card ignores. Can be set to `.top` to allow a card to affect status bar area (i.e. let an image or color reach the top of the screen). Defaults to `nil`.
public var cardIgnoreSafeAreaEdges: Edge.Set?

/// The padding around the bottom controls. Defaults to `.init(top: 0, leading: 24, bottom: 24, trailing: 24)`.
public var bottomControlsPadding: EdgeInsets

/// The spacing between the bottom controls. Defaults to `16`.
public var bottomControlsSpacing: CGFloat

/// The height of the navigation buttons. May be used to negate visual glitches if navigation button font size and icon size differ too much, if `nil`, defaults to intrinsic content size.
public var navigationButtonHeight: CGFloat?

/// The padding around the navigation button content. Defaults to `init(top: 16, leading: 16, bottom: 16, trailing: 16)`.
public var navigationButtonContentPadding: EdgeInsets

/// The squared size of navigation button icons, i.e. `xmark`, `arrow.left` and `arrow.right`. Defaults to `20`.
public var navigationButtonIconSize: CGFloat

/// The dismiss icon used in the navigation buttons. Defaults to `xmark`.
public var navigationButtonDismissIcon: Image

/// The previous icon used in the navigation buttons. Defaults to `arrow.left`.
public var navigationButtonPreviousIcon: Image

/// The next icon used in the navigation buttons. Defaults to `arrow.right`.
public var navigationButtonNextIcon: Image

/// The finish text used in the right navigation button on the last page. Defaults to `Let's start`.
public var navigationButtonFinishText: String

/// If `true`, shows navigation buttons at the bottom of the cards carousel. If `false`, hides the navigation buttons.
public var showNavigationButtons: Bool

/// If `true`, shows indicators at the bottom of the cards carousel. If `false`, hides the indicators.
public var showIndicators: Bool

/// If `true`, will allow the user to swipe the cards left and right, if `false`, prevents any swipe interaction.
public var cardsSwipingEnabled: Bool

/// The tint color of all carousel elements if not overidden. Including primary button background color, secondary button text color and indicator color.
public var tintColor: Color

/// The backgroundColor of the carousel view, i.e. the area arround the card.
public var backgroundColor: Color

/// The text color of content within the primary buttons, i.e. the right navigation button. Defaults to `.white`.
public var primaryButtonForegroundColor: Color

/// The background color of primary buttons, i.e. the right navigation button. Defaults to `tintColor`.
public var primaryButtonBackgroundColor: Color?

/// The text color of content within the secondary buttons, i.e. the left navigation button. If `nil`, defaults to `tintColor`.
public var secondaryButtonForegroundColor: Color?

/// The background color of primary buttons, i.e. the left navigation button. Defaults to `primaryButtonTextColor`, which defaults to `white`.
public var secondaryButtonBackgroundColor: Color?

/// The color of active indicators. If `nil`, defaults to `tintColor`.
public var activeIndicatorColor: Color?

/// The color of inactive indicators. If `nil`, defaults to `tintColor` with `0.5 opacity`.
public var inactiveIndicatorColor: Color?

/// The font used for all button texts, i.e. the skip and continue button below the center card.
public var buttonTextFont: Font

/// The corner radius of the card. Defaults to `24`.
public var cardCornerRadius: CGFloat

/// The corner radius of the buttons. Defaults to `16`.
public var buttonCornerRadius: CGFloat

/// The corner radius of the indicators. Defaults to `.infinity` to make them round.
public var indicatorCornerRadius: CGFloat

/// The spacing between indicators. Defaults to `6`.
public var indicatorSpacing: CGFloat

/// The size of each indicator, determines both width and height of circle. Defaults to `6`.
public var indicatorSize: CGFloat

/// The animation of the entire carousel. Defaults to `.default`, but replaced with `nil` during onAppear to prevent visual glitches in `nonSwipeableCards`.
public var carouselAnimation: Animation?
```

## 🔗 Related publications

Look at our other repositories on our [GitHub account](https://github.com/orgs/Coffee-IT-Development/repositories).

## ✏️ Changelog

All notable changes to this project will be documented in the [Changelog](CHANGELOG.md).
`CITCardsCarousel` adheres to [Semantic Versioning](https://semver.org/).

## 📧 Contact

For questions, ideas or help you can reach us by email at contact@coffeeit.nl.

## ⚠️ License

Distributed under the MIT License. [See LICENSE](LICENSE.md) for more information.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
`CITCardsCarousel` adheres to [Semantic Versioning](https://semver.org/).

#### 1.x Releases
- `1.0.x` Releases - [1.0.0](#100) | [1.0.1](#101)

#### 0.x Releases
- `0.1.x` Releases - [0.1.0](#010)

#### 1.0.1

- Updated README to new standard.

#### 1.0.0

- Add more configuration options, improve documentation, add CITCardsCarouselView preview.

#### 0.1.0

- Setup CITCardsCarousel, add customization & documentation.


Sources/CITCardsCarousel/CITCardsCarouselConfiguration.swift
//
//  CITCardsCarouselConfiguration.swift
//
//  MIT License
//
//  Copyright (c) 2022 Coffee IT
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.
//

import SwiftUI

/// Normally we add Configuration as an extension to the core view, like CITCardsCarouselView+Configuration, however, due to CITCardsCarouselView's generic constraints, that would complicate internal code.
public struct CITCardsCarouselConfiguration: Equatable {
    
    /// The padding around the card container. Defaults to `.init(top: 24, leading: 24, bottom: 32, trailing: 24)`.
    public var cardPadding: EdgeInsets
    
    /// The safe area edges that a card ignores. Can be set to `.top` to allow a card to affect status bar area (i.e. let an image or color reach the top of the screen). Defaults to `nil`.
    public var cardIgnoreSafeAreaEdges: Edge.Set?
    
    /// The padding around the bottom controls. Defaults to `.init(top: 0, leading: 24, bottom: 24, trailing: 24)`.
    public var bottomControlsPadding: EdgeInsets
    
    /// The spacing between the bottom controls. Defaults to `16`.
    public var bottomControlsSpacing: CGFloat
    
    /// The height of the navigation buttons. May be used to negate visual glitches if navigation button font size and icon size differ too much, if `nil`, defaults to intrinsic content size.
    public var navigationButtonHeight: CGFloat?
    
    /// The padding around the navigation button content. Defaults to `init(top: 16, leading: 16, bottom: 16, trailing: 16)`.
    public var navigationButtonContentPadding: EdgeInsets
    
    /// The squared size of navigation button icons, i.e. `xmark`, `arrow.left` and `arrow.right`. Defaults to `20`.
    public var navigationButtonIconSize: CGFloat
    
    /// The dismiss icon used in the navigation buttons. Defaults to `xmark`.
    public var navigationButtonDismissIcon: Image
    
    /// The previous icon used in the navigation buttons. Defaults to `arrow.left`.
    public var navigationButtonPreviousIcon: Image
    
    /// The next icon used in the navigation buttons. Defaults to `arrow.right`.
    public var navigationButtonNextIcon: Image
    
    /// The finish text used in the right navigation button on the last page. Defaults to `Let's start`.
    public var navigationButtonFinishText: String
    
    /// If `true`, shows navigation buttons at the bottom of the cards carousel. If `false`, hides the navigation buttons.
    public var showNavigationButtons: Bool
    
    /// If `true`, shows indicators at the bottom of the cards carousel. If `false`, hides the indicators.
    public var showIndicators: Bool
    
    /// If `true`, will allow the user to swipe the cards left and right, if `false`, prevents any swipe interaction.
    public var cardsSwipingEnabled: Bool
    
    /// The tint color of all carousel elements if not overidden. Including primary button background color, secondary button text color and indicator color.
    public var tintColor: Color
    
    /// The backgroundColor of the carousel view, i.e. the area arround the card.
    public var backgroundColor: Color
    
    /// The text color of content within the primary buttons, i.e. the right navigation button. Defaults to `.white`.
    public var primaryButtonForegroundColor: Color
    
    /// The background color of primary buttons, i.e. the right navigation button. Defaults to `tintColor`.
    public var primaryButtonBackgroundColor: Color?
    
    /// The text color of content within the secondary buttons, i.e. the left navigation button. If `nil`, defaults to `tintColor`.
    public var secondaryButtonForegroundColor: Color?
    
    /// The background color of primary buttons, i.e. the left navigation button. Defaults to `primaryButtonTextColor`, which defaults to `white`.
    public var secondaryButtonBackgroundColor: Color?
    
    /// The color of active indicators. If `nil`, defaults to `tintColor`.
    public var activeIndicatorColor: Color?
    
    /// The color of inactive indicators. If `nil`, defaults to `tintColor` with `0.5 opacity`.
    public var inactiveIndicatorColor: Color?
    
    /// The font used for all button texts, i.e. the skip and continue button below the center card.
    public var buttonTextFont: Font
    
    /// The corner radius of the card. Defaults to `24`.
    public var cardCornerRadius: CGFloat
    
    /// The corner radius of the buttons. Defaults to `16`.
    public var buttonCornerRadius: CGFloat
    
    /// The corner radius of the indicators. Defaults to `.infinity` to make them round.
    public var indicatorCornerRadius: CGFloat
    
    /// The spacing between indicators. Defaults to `6`.
    public var indicatorSpacing: CGFloat
    
    /// The size of each indicator, determines both width and height of circle. Defaults to `6`.
    public var indicatorSize: CGFloat
    
    /// The animation of the entire carousel. Defaults to `.default`, but replaced with `nil` during onAppear to prevent visual glitches in `nonSwipeableCards`.
    public var carouselAnimation: Animation?
    
    public init(
        cardPadding: EdgeInsets                         = .init(top: 24, leading: 24, bottom: 32, trailing: 24),
        cardIgnoreSafeAreaEdges: Edge.Set?              = nil,
        bottomControlsPadding: EdgeInsets               = .init(top: 0, leading: 24, bottom: 24, trailing: 24),
        bottomControlsSpacing: CGFloat                  = 16,
        navigationButtonHeight: CGFloat?                = nil,
        navigationButtonContentPadding: EdgeInsets      = .init(top: 16, leading: 16, bottom: 16, trailing: 16),
        navigationButtonIconSize: CGFloat               = 20,
        navigationButtonDismissIcon: Image              = .init(systemName: "xmark"),
        navigationButtonPreviousIcon: Image             = .init(systemName: "arrow.left"),
        navigationButtonNextIcon: Image                 = .init(systemName: "arrow.right"),
        navigationButtonFinishText: String              = "Let's start",
        showNavigationButtons: Bool                     = true,
        showIndicators: Bool                            = true,
        cardsSwipingEnabled: Bool                       = true,
        tintColor: Color                                = .blue,
        backgroundColor: Color                          = .gray,
        primaryButtonForegroundColor: Color             = .white,
        primaryButtonBackgroundColor: Color?            = nil,
        secondaryButtonForegroundColor: Color?          = nil,
        secondaryButtonBackgroundColor: Color?          = nil,
        activeIndicatorColor: Color?                    = nil,
        inactiveIndicatorColor: Color?                  = nil,
        buttonTextFont: Font                            = .system(size: 16),
        cardCornerRadius: CGFloat                       = 24,
        buttonCornerRadius: CGFloat                     = 16,
        indicatorCornerRadius: CGFloat                  = .infinity,
        indicatorSpacing: CGFloat                       = 6,
        indicatorSize: CGFloat                          = 6,
        carouselAnimation: Animation?                   = .default
    ) {
        self.cardPadding = cardPadding
        self.cardIgnoreSafeAreaEdges = cardIgnoreSafeAreaEdges
        self.bottomControlsPadding = bottomControlsPadding
        self.bottomControlsSpacing = bottomControlsSpacing
        self.navigationButtonHeight = navigationButtonHeight
        self.navigationButtonContentPadding = navigationButtonContentPadding
        self.navigationButtonIconSize = navigationButtonIconSize
        self.navigationButtonDismissIcon = navigationButtonDismissIcon
        self.navigationButtonPreviousIcon = navigationButtonPreviousIcon
        self.navigationButtonNextIcon = navigationButtonNextIcon
        self.navigationButtonFinishText = navigationButtonFinishText
        self.showNavigationButtons = showNavigationButtons
        self.showIndicators = showIndicators
        self.cardsSwipingEnabled = cardsSwipingEnabled
        self.tintColor = tintColor
        self.backgroundColor = backgroundColor
        self.primaryButtonForegroundColor = primaryButtonForegroundColor
        self.primaryButtonBackgroundColor = primaryButtonBackgroundColor ?? tintColor
        self.secondaryButtonForegroundColor = secondaryButtonForegroundColor ?? tintColor
        self.secondaryButtonBackgroundColor = secondaryButtonBackgroundColor ?? primaryButtonForegroundColor
        self.activeIndicatorColor = activeIndicatorColor ?? tintColor
        self.inactiveIndicatorColor = inactiveIndicatorColor ?? tintColor.opacity(0.5)
        self.buttonTextFont = buttonTextFont
        self.cardCornerRadius = cardCornerRadius
        self.buttonCornerRadius = buttonCornerRadius
        self.indicatorCornerRadius = indicatorCornerRadius
        self.indicatorSpacing = indicatorSpacing
        self.indicatorSize = indicatorSize
        self.carouselAnimation = carouselAnimation
    }
    
    public static var example = CITCardsCarouselConfiguration()
}
